presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSink.java

```diff
@@ -0,0 +1,491 @@
+                    context.getWriter().rollback();
+                }
+            }
+            catch (Throwable t) {
+                if (error == null) {
+                    error = new RuntimeException("Exception during rollback");
+                }
+                error.addSuppressed(t);
+            }
+        }
+        if (error != null) {
+            throw error;
+        }
+    }
+
+    private void doAppend(Page page)
+    {
+        while (page.getPositionCount() > MAX_PAGE_POSITIONS) {
+            Page chunk = page.getRegion(0, MAX_PAGE_POSITIONS);
+            page = page.getRegion(MAX_PAGE_POSITIONS, page.getPositionCount() - MAX_PAGE_POSITIONS);
+            writePage(chunk);
+        }
+
+        writePage(page);
+    }
+
+    private void writePage(Page page)
+    {
+        int[] writerIndexes = getWriterIndexes(page);
+
+        // position count for each writer
+        int[] sizes = new int[writers.size()];
+        for (int index : writerIndexes) {
+            sizes[index]++;
+        }
+
+        // record which positions are used by which writer
+        int[][] writerPositions = new int[writers.size()][];
+        int[] counts = new int[writers.size()];
+
+        for (int position = 0; position < page.getPositionCount(); position++) {
+            int index = writerIndexes[position];
+
+            int count = counts[index];
+            if (count == 0) {
+                writerPositions[index] = new int[sizes[index]];
+            }
+            writerPositions[index][count] = position;
+            counts[index]++;
+        }
+
+        // invoke the writers
+        for (int index = 0; index < writerPositions.length; index++) {
+            int[] positions = writerPositions[index];
+            if (positions == null) {
+                continue;
+            }
+
+            // if write is partitioned across multiple writers, filter page using dictionary blocks
+            Page pageForWriter = page;
+            if (positions.length != page.getPositionCount()) {
+                verify(positions.length == counts[index]);
+                pageForWriter = pageForWriter.getPositions(positions, 0, positions.length);
+            }
+
+            IcebergFileWriter writer = writers.get(index).getWriter();
+
+            long currentWritten = writer.getWrittenBytes();
+            long currentMemory = writer.getSystemMemoryUsage();
+
+            writer.appendRows(pageForWriter);
+
+            writtenBytes += (writer.getWrittenBytes() - currentWritten);
+            systemMemoryUsage += (writer.getSystemMemoryUsage() - currentMemory);
+        }
+    }
+
+    private int[] getWriterIndexes(Page page)
+    {
+        int[] writerIndexes = pagePartitioner.partitionPage(page);
+
+        if (pagePartitioner.getMaxIndex() >= maxOpenWriters) {
+            throw new PrestoException(ICEBERG_TOO_MANY_OPEN_PARTITIONS, format("Exceeded limit of %s open writers for partitions", maxOpenWriters));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: maxOpenWriters. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 267. Consider formatting this data with &#39;%d&#39; instead.